core/state/trie_prefetcher.go

```diff
@@ -281,30 +278,20 @@ func newSubfetcher(db Database, state common.Hash, owner common.Hash, root commo
 
 // schedule adds a batch of trie keys to the queue to prefetch.
 func (sf *subfetcher) schedule(addrs []common.Address, slots []common.Hash, read bool) error {
-	// Ensure the subfetcher is still alive
-	select {
-	case <-sf.term:
-		return errTerminated
-	default:
-	}
-	// Append the tasks to the current queue
-	sf.lock.Lock()
+	tasks := make([]*subfetcherTask, 0, len(addrs)+len(slots))
 	for _, addr := range addrs {
-		sf.tasks = append(sf.tasks, &subfetcherTask{read: read, addr: &addr})
+		tasks = append(tasks, &subfetcherTask{read: read, addr: &addr})
 	}
 	for _, slot := range slots {
-		sf.tasks = append(sf.tasks, &subfetcherTask{read: read, slot: &slot})
+		tasks = append(tasks, &subfetcherTask{read: read, slot: &slot})
 	}
-	sf.lock.Unlock()
 
-	// Notify the background thread to execute scheduled tasks
 	select {
-	case sf.wake <- struct{}{}:
-		// Wake signal sent
-	default:
-		// Wake signal not sent as a previous one is already queued
+	case sf.tasks <- tasks:
```

Comment:
This is now blocking. Previously, any number of threads could append to the `sf.tasks`, and they could all exit afterwards, since they select was not blocking. 

Now, instead, they all allocate their own slice of tasks, and each thread need to wait until it has been delivered over the `sf.tasks` channel. 

This seems to be quite a bit change in semantics, yet not something you mentioned on the PR description. I'm guessing it was unintentional. Unless I have misunderstood something, I think this highlights why we should reject this PR: it touches some very intricate parts of geth, which currently works fine. It's very easy to slip when making changes here.

Reply:
I've [responsed](https://github.com/ethereum/go-ethereum/pull/30782#issuecomment-2503614237) to this in main conversation thread.